### PR TITLE
Incorporated review comments and added some exception handling resource cleanup

### PR DESCRIPTION
**Pull request by dservgun**
@guoshimin I have added some more changes to incorporate your comments.  

Tasks that i still need to do: 
. Actually setup a client and get the exec helper to work.
. Should i be using katip logging? 
. Should i be adding a signal handler as well to handle sigint? 

``` TimeoutInterval``` , could probably be replaced with [Data.Time.Interval](http://hackage.haskell.org/package/time-interval-0.1.1/docs/Data-Time-Interval.html#t:TimeInterval).

### CODE REVIEW

**Comment by guoshimin** [comment]
Thanks @dservgun
 
- I'm not familiar with the logging landscape in Haskell. I think it's fine if the initial version has no logging.
- I don't think you need to have an SIGINT handler. This library is meant to be embedded in other applications that want to execute commands in a remote Kubernetes pod. How to handle signals is the concern of those other applications. I think here we want to expose an interface similar to createProcess in System.Process.

File: kubernetes-exec-helper/src/Kubernetes/WSClient.hs [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+-- | Read the text from the channels and direct to the appropriate 
+-- | local channel.
+writeToLocalChannels :: [(ChannelId, TChan Text)] -> IO [Async ThreadId]
+writeToLocalChannels channels = 
+  mapM (flip writeToLocalChannel channels) $ [StdIn, StdOut, StdErr] 
+
+-- | Write to a local channel. 
+-- | === Note : The word local channel is used to represent the 
+-- | command line from which a user is running a command.
+writeToLocalChannel :: ChannelId -> [(ChannelId, TChan Text)] -> IO (Async ThreadId)
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/WSClient.hs]: Change the order of the first two parameters so you don't need to `flip` in `writeToLocalChannels`?

File: kubernetes-exec-helper/src/Kubernetes/WSClient.hs [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+writeToLocalChannel channelId channels = do 
+  let std = mapChannel channelId
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/WSClient.hs]: I don't think we should connect these streams directly to the calling process's stdin/stdout/stderr. It should be up to the calling code what to do with the data coming from these streams.

File: kubernetes-exec-helper/src/Kubernetes/WSClient.hs [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+  hSetBuffering std NoBuffering
+  async $ forever $ do
+        message <- atomically $ readTChan $ getTChanSTM channelId channels
+        T.hPutStr std message
+
+-- | Send 'stdin' to the pod.
+readFromStdIn :: TChan Text -> IO (Async ThreadId)
+readFromStdIn outputChan = do 
+  hSetBuffering stdin NoBuffering
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/WSClient.hs]: Again, it should be up to the calling code how to supply data to the stdin for the remote exec.

**Comment by dservgun** [comment]
@guoshimin, I moved the code around into a new module WSStream, as in the reference implementation. I will probably try to expose a ``` createClient``` interface in the Stream module. I will update this pr with that change. ``` newtype ClientState a = ClientState {
    _unState :: 
      ([Async ThreadId], TChan a, [(ChannelId, TChan a)])
    }
```.  If i expose this newtype as part of the createClient call, will that suffice?

**Comment by dservgun** [comment]
@guoshimin A question about kubeconfig, I was trying this yaml file: 
``` spec: 
  containers:
    image : busybox
    name : sleep
    args : 
      /bin/sh
      -c 
      while true; do date sleep 5 done

```
I dont see the parser picking up these elements. If i need to add this to the kubeconfig, how do I go about it? I can try to put together some initial version for your review. I am trying to reproduce exec.py. 
Thanks.

**Comment by dservgun** [comment]
@guoshimin : thank you for the earlier comments. I have tried to rename ``` ClientState ``` with ```CreateWSClient ``` to reflect the intent of ``` CreateProcess in System.Process ```. I was also trying to parse the example "busybox.yaml" file, I had to add a few elements in KubeConfig. Can you please take a look and let me know if that works.

File: kubeconfig/src/Kubernetes/KubeConfig.hs [hunk 4]
```diff
@@ -48,6 +48,7 @@ data Config = Config
   , authInfos      :: [NamedEntity AuthInfo "user"]
   , contexts       :: [NamedEntity Context "context"]
   , currentContext :: Text
+  , spec           :: Spec
```
Comment (guoshimin) [inline kubeconfig/src/Kubernetes/KubeConfig.hs]: This should not be here. This is the type of a kubeconfig, whose format is defined here: https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/api/v1/types.go

File: kubeconfig/src/Kubernetes/KubeConfig.hs [hunk 5]
```diff
@@ -89,14 +90,34 @@ data NamedEntity a (typeKey :: Symbol) = NamedEntity
...
 
 instance (ToJSON a, KnownSymbol s) =>
          ToJSON (NamedEntity a s) where
   toJSON (NamedEntity {..}) = object
       ["name" .= toJSON name, T.pack (symbolVal (Proxy :: Proxy s)) .= toJSON entity]
 
+newtype Spec = Spec {containers :: [Container]}
+  deriving (Eq, Generic, Show, Typeable)
+
+data Container = Container {
```
Comment (guoshimin) [inline kubeconfig/src/Kubernetes/KubeConfig.hs]: You don't need to define a Container type. Use V1Container in https://raw.githubusercontent.com/kubernetes-client/haskell/master/kubernetes/lib/Kubernetes/Model.hs.

File: kubernetes-exec-helper/examples/Exec.hs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                  manager
+                  config
+                  request 
+                    >>= print 
+    Left e -> print e 
+  return ()
+
+main :: IO ()
+main = do 
+  exec =<< execParser opts
```
Comment (guoshimin) [inline kubernetes-exec-helper/examples/Exec.hs]: Where is execParser defined?

File: kubernetes-exec-helper/src/Kubernetes/WSClient.hs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    Just aChan -> 
+      atomically $
+        writeTChan (snd $ getChannelIdSTM aChan channels) message
+
+{- | 
+  Query a channelId from a list of Channels.
+-}
+getTChanSTM :: ChannelId -> [(ChannelId, TChan Text)] -> TChan Text 
+getTChanSTM a b = snd $ getChannelIdSTM a b 
+getChannelIdSTM :: ChannelId -> [(ChannelId, TChan Text)] -> (ChannelId, TChan Text)
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/WSClient.hs]: Why not just return `TChan Text`?

File: kubernetes-exec-helper/src/Kubernetes/WSClient.hs [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+getChannelIdSTM aChannelId channels = 
+  head $ Prelude.filter(\(x, _) -> x == aChannelId) channels
+
+readChannelIdSTM :: ChannelId -> [(ChannelId, TChan Text)] -> STM Text
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/WSClient.hs]: What's the purpose of these read/write functions? I don't see them used anywhere.

File: kubernetes-exec-helper/src/Kubernetes/K8SChannel.hs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+newtype InvalidChannel = InvalidChannel Text deriving (Show, Typeable)
+
+
+{- | 
+  CreateWSClient maintains all threads that are running,  
+  a writer channel to send messages to the server 
+  and a list of all "(ChannelId, TChan Text)" pairs.
+  Clients can wait on '[Async ThreadId]' and proceed to work with each channel.
+-}
+newtype CreateWSClient a = CreateWSClient {
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/K8SChannel.hs]: This is not really analogous to how System.Process works. In System.Process, the [CreateProcess](https://hackage.haskell.org/package/process-1.6.3.0/docs/System-Process.html#t:CreateProcess) structure contains information about the how to run the process: the executable, the arguments, the environment variables, etc, and the [createProcess](https://hackage.haskell.org/package/process-1.6.3.0/docs/System-Process.html#v:createProcess) function returns handles for manipulating stdin/stdout/stderr as well as the spawned process itself. The equivalent of CreateProcess here would be a structure that defines the command to execute and what container of what pod to execute that command in, and the equivalent of createProcess will return channels for reading the stdout/stderr of the command and sending data to stdin, as well as a handle for managing the websocket session itself.

File: kubernetes-exec-helper/src/Kubernetes/K8SChannel.hs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+-- | Command contains the "Executable" and a list of "Arguments"
+type Command = String
+
+-- | A reader configuration when running the client.
+data ExecClientConfig = 
+  ExecClientConfig {
+  _kubeConfig :: KubeConfig
+  , _url :: URL 
+  , _timeout :: Maybe TimeoutInterval
+  , _preload :: Bool
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/K8SChannel.hs]: Looks like this field is irrelevant.

**Comment by dservgun** [comment]
@guoshimin, i am still working on changing a bunch of files. Please try to connect with me on slack before you review the changes. Or better still, should I pull my pr and resubmit? Thanks.

**Comment by dservgun** [comment]
Hi @mbohlool @guoshimin i am getting bad request error from the test exec. I have wire shark output from the python reference implementation and my changes. Will you have some time to help me debug this please so I can get unstuck. Thanks

File: kubernetes-exec-helper/src/Kubernetes/CreateWSClient.hs [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+clusterClientParams client = _clusterClientParams client 
+
+getTimeOut :: CreateWSClient a -> Maybe TimeoutInterval 
+getTimeOut client = _timeOutInterval client
+
+{-| 
+  A convenience method to return a client state with relevant 
+  reader and writer channels.
+-}
+createWSClient :: TLS.ClientParams -> KubernetesConfig -> V1Container -> [Command] -> IO (CreateWSClient Text)
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/CreateWSClient.hs]: This interface feels off to me. If you created this interface at my previous suggestion of an interface similar to System.Process.createProcess, then this doesn't really capture the spirit of the latter. The spirit of the latter is that you use a pure structure (CreateProcess) that fully specifies how to invoke a command, and createProcess takes that pure structure and returns a set of objects in an IO monad that lets you interact with the spawned process.

To a first approximation, the interface here should look something like
```
createWSClient :: CreateWSClient -> IO (TChan, TChan, TChan, WSClient)
```
It should make the API call and start the websocket session. The three returned channels can be used to send input to the remote process and read the stdout/stderr of the remote process. The returned WSClient can be used to terminate the session or retrieve the exit status.

File: kubernetes-exec-helper/src/Kubernetes/WSClient.hs [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+import System.IO (hSetBuffering, BufferMode(..), stdin)
+import Wuss as WSS (runSecureClient)
+
+
+runClient :: CreateWSClient Text -> Name -> Namespace -> AuthApiKeyBearerToken -> IO () 
+runClient createWSClient name namespace bearerToken = do 
+  let 
+    headers = getHeaders createWSClient
+    connectionOptions = getConnectionOptions createWSClient
+    uri  = getURIAuth createWSClient
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/WSClient.hs]: Where is uri used?

File: kubernetes-exec-helper/src/Kubernetes/CreateWSClient.hs [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+kubernetesConfig = _kubernetesConfig
+
+commands :: CreateWSClient a -> [Command]
+commands client = _commands client
+
+clusterClientParams :: CreateWSClient a -> TLS.ClientParams 
+clusterClientParams client = _clusterClientParams client 
+
+getTimeOut :: CreateWSClient a -> Maybe TimeoutInterval 
+getTimeOut client = _timeOutInterval client
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/CreateWSClient.hs]: All these one-line functions seem quite unnecessary to me. Am I missing something?

File: kubernetes-exec-helper/src/Kubernetes/K8SChannel.hs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+-- | Decode text to a channel.
+readChannel :: Text -> Maybe ChannelId
+readChannel "0" = Just StdIn
+readChannel "1" = Just StdOut 
+readChannel "2" = Just StdErr 
+readChannel "3" = Just Error 
+readChannel "4" = Just Resize 
+readChannel _ = Nothing
+
+-- | Tie the channel back to the handles on the command line.
```
Comment (guoshimin) [inline kubernetes-exec-helper/src/Kubernetes/K8SChannel.hs]: As I said in another comment, this library should not deal with the calling process' stdin/stdout/stderr at all.

**Comment by guoshimin** [comment]
@dservgun Oh sorry saw your message about connecting with you on slack after having already left comments. Will try to find you on slack.

File: kubernetes-client-helper/src/Kubernetes/ClientHelper.hs [hunk 15]
```diff
@@ -44,6 +44,13 @@ setMasterURI server kcfg =
 disableValidateAuthMethods :: K.KubernetesConfig -> K.KubernetesConfig
 disableValidateAuthMethods kcfg = kcfg { K.configValidateAuthMethods = False }
 
+-- | Configures the 'K.KubernetesConfig' to using token authentication
```
Comment (dservgun) [inline kubernetes-client-helper/src/Kubernetes/ClientHelper.hs]: a convenience method so I could handle bearertoken.

File: kubernetes-exec-helper/src/Kubernetes/WSClient.hs [hunk 16]
```diff
@@ -60,50 +63,51 @@ import qualified Network.WebSockets.Stream as WS
...
+  runClientWithTLS (host req) (port req) (endpoint req) kubeConfig clientParams (\conn -> k8sClient timeoutInt createWSClient conn)
+  wait execAttach_ >> return ()
+  where 
+    endpoint req = 
+      T.unpack $ 
+        T.pack $
+          BC.unpack $
+            NH.method req <> " " <> NH.host req <> NH.path req <> NH.queryString req
+    host req = T.unpack $ T.pack $ BC.unpack $ NH.host req 
+    port req = PortNum 8443 -- $ ((fromIntegral (NH.port req :: Int)))
```
Comment (dservgun) [inline kubernetes-exec-helper/src/Kubernetes/WSClient.hs]: Need to pick up the value from the kube.

**Comment by dservgun** [comment]
@guoshimin : I guess i was able to fix the core connection issues so now the client example code connects to the pod and outputs on stdout. I still need to do some more testing. I will continue to make some changes. There are some functions that I wish I can write better. Let me add a comment at relevant places. Thanks.

**Comment by dservgun** [comment]
@guoshimin I may have completely missed your messages on slack. I just saw them. Agree to all the points. Lets try to meet sometime today. Current state is that the example code can connect to minikube and run the ``` date ``` command. I need to figure out as to how to pass parameters to the command.

**Pull request closed without merging**